### ISSUE TITLE
Qwen3Next / gated-delta architectures fall back to sequential decode

### ISSUE THREAD

**Issue by TheTom** [TheTom/vllm-swift#8]
## Bug

Models in the Qwen3Next family (Qwen3.5, Qwen3.6, gated-delta hybrid architectures) do not use the fully batched concurrent decode path. They fall back to per-request sequential decode, so total throughput at B>1 does not scale with concurrency.

Signature: `B=1`, `B=8`, `B=32`, `B=64` all return roughly the same total tok/s.

Documented in README under Known Limitations:

> Only Qwen3 models use the fully batched decode path; other architectures fall back to sequential decode (still functional, just slower at high concurrency)

Filing as a tracked issue so progress is visible.

## Repro

```bash
vllm-swift serve ~/models/Qwen3.6-27B-UD-MLX-4bit --max-model-len 4096
# Test at B=1, B=8, B=32, B=64. Total tok/s flat across concurrency levels.
```

Observed example (Chander, M4 Max 64GB, Qwen3.6-27B-UD-MLX-4bit):
```
B=1:  18.6 tok/s
B=8:  19.0 tok/s total
B=32: 19.1 tok/s total
B=64: 19.1 tok/s total
```

vs Qwen3-4B-4bit on the same hardware where total throughput scales meaningfully across B.

## Fix scope

The Qwen3Next architecture is hybrid: alternating GatedDeltaNet (recurrent state) and attention layers. The current `BatchedKVCache` only models attention layers, so hybrid models fall back to per-request iteration through the model.

Two pieces:

1. **Batched cache for hybrid models.** New `BatchedHybridCache` (or extension of `BatchedKVCache`) that holds per-slot attention KV and per-slot GatedDeltaNet recurrent state. Lifecycle compatible with existing bridge entry points (`vsm_engine_init_batched`, `vsm_engine_add_batch_slot`, `vsm_engine_remove_batch_slot`).

2. **Batched forward for Qwen3Next.** Replace the per-request iteration with a true batched forward that:
   - Routes attention layers through the `BatchedKVCache` path
   - Routes GatedDeltaNet layers through a batched recurrent step (state shape `[B, state_dim]`, single kernel dispatch per layer)

The GDN kernel `gated_delta_step_fused_*` already exists but needs verification that it accepts `B > 1` input shape, or needs extension to do so.

## Out of scope

- Architecture changes to GatedDeltaNet itself
- Memory layout changes to the existing `BatchedKVCache` (extend with composition, don't replace)
- Other architectures (MoE handled separately, MLA / DeepSeek separately)

## Related

- README Known Limitations
- Reported via community testing of Qwen3.6-27B on M4 Max

**Comment by TheTom** [comment]
Will follow up on the X thread once fixed: https://x.com/no_stp_on_snek/status/2048384960751407438

**Comment by TheTom** [comment]
## Fixed locally — pending PR

Root cause: GDN Metal kernel (`gated_delta_step` / `gated_delta_step_fused`) uses raw pointer arithmetic with assumed contiguous strides:

```cpp
auto q_ = q + b_idx * T_val * Hk * Dk + hk_idx * Dk;
```

But `MLX.split(convOut, indices: [keyDim, 2*keyDim], axis: -1)` returns non-contiguous slice views with stride `convDim` along axis 1, not `keyDim`/`valueDim`. At B=1 only `b_idx=0` is used so the stride mismatch is invisible. At B>1 the kernel reads into the wrong slot's region of memory → garbage tokens for slots 1+.

Fix: `.contiguous()` on q/k/v after the split + reshape, in both `Qwen35GatedDeltaNet` and `Qwen3NextGatedDeltaNet` (`callAsFunction` + `fullyBatchedForward` in each). Surgical 4-line change per file.

## Proof

### Regression test bisects the bug exactly

`Qwen35BatchedHybridCacheTests.fullyBatchedDecode produces identical outputs for identical inputs across slots` — same input token to every slot at B=4, asserts max abs logit diff per slot vs slot 0 == 0.

**With fix:**
```
✔ Test "fullyBatchedDecode produces identical outputs for identical inputs across slots" passed
```

**Without fix (deliberately reverted to bisect):**
```
✘ slot 1 logits diverged from slot 0 by max abs 0.046875
✘ slot 2 logits diverged from slot 0 by max abs 0.037109375
✘ slot 3 logits diverged from slot 0 by max abs 0.015625
```

Test reliably catches the bug. Re-applied fix passes immediately after.

### User repro fixed (Qwen3.6-27B-UD-MLX-4bit, M5 Max 128GB)

| B | Reported (broken) | Now | Scaling |
|---|---:|---:|---:|
| 1 | 18.6 | 20.1 | 1.0× |
| 8 | 19.0 | 63.8 | **3.2×** |
| 32 | 19.1 | 176.3 | **9×** |
| 64 | 19.1 | **305.6** | **16×** |

Decode throughput now scales with B as expected. Prior flat ~19 tok/s pattern across B is gone.

### Hybrid sweep — all scale

| Model | B=1 decode | B=8 | B=32 | B=64 |
|---|---:|---:|---:|---:|
| Qwen3.5-2B-4bit | 280 | 828 | 1,549 | **2,029** |
| Qwen3.5-9B-4bit | 92 | 261 | 542 | **781** |
| Qwen3.5-35B-A3B-4bit (MoE) | 134 | 369 | 616 | **994** |
| Qwen3.6-27B-UD-MLX-4bit | 20 | 64 | 176 | **306** |

MoE variant (Qwen3.5-35B-A3B) verifies the inherited `BatchedHybridLLM` conformance works through MoE expert routing.

### Dense Qwen3 regression — zero regression

| Cell | Result | Baseline | Δ |
|---|---:|---:|---:|
| Qwen3-4B p18 B=64 | 1,513 | 1,492 | +1.4% |
| Qwen3-0.6B p18 B=64 | 3,362 | 3,282 | +2.4% |
| Qwen3-4B p2048 B=64 unique | 770 | 760 | +1.3% |
| Qwen3-4B p8192 B=64 unique | 273 | 269 | +1.5%, no OOM |

All within ±5%. Qwen3 dense fast path bit-identical (cast first, hybrid second).

## Commits (local only — not yet pushed)

- mlx-swift-lm `feat/paged-attention`: `07e9e60`, `1ce5144`, `ac8b796`, `1ea596b`, `7a60b57`, `c6358a6`, `099dffc`, `3cf0b94`
- vllm-swift `feat/paged-attention`: `dfc5bae`, `419be2b`, `a791c37`

Will leave open until both PRs are up and merged.

**Comment by TheTom** [comment]
## Mac Mini M2 Pro verification (32 GB, macOS 26.1)

Tested on a lesser device to confirm no architecture-specific regression and that the fix holds outside M5 Max.

**Swift unit test** — `Qwen35BatchedHybridCacheTests.fullyBatchedDecode produces identical outputs for identical inputs across slots` runs the same bisect oracle as on M5 Max:

\`\`\`
✔ Suite "BatchedHybridCache + BatchedMambaCache" passed after 0.587 seconds.
✔ Suite "PagedKVCache foundation" passed after 0.029 seconds.
◇ Test "fullyBatchedDecode produces identical outputs for identical inputs across slots" started.
✔ Test "fullyBatchedDecode produces identical outputs for identical inputs across slots" passed after 0.997 seconds.
✔ Suite "Qwen35Model BatchedHybridLLM cache lifecycle" passed after 1.011 seconds.
✔ Suite "Qwen35MoEModel BatchedHybridLLM cache lifecycle" passed after 0.010 seconds.
\`\`\`

**Hybrid scaling — Qwen3.5-2B-4bit** (decode tok/s, prompt=18, identical, sequential prefill):

| B | decode tok/s | scaling vs B=1 |
|---|---:|---:|
| 1 | 107 | 1.0× |
| 8 | 123 | 1.15× |
| 32 | 329 | **3.1×** |
| 64 | 368 | **3.4×** |

Pre-fix would have been flat ~107 across all B (same flatness pattern as Qwen3.6-27B on M4 Max in the original bug report). M2 Pro scaling is naturally lower than M5 Max (which hit 7.2× on the same model) due to less compute headroom and earlier memory-bandwidth saturation, but the bug is fully resolved.

**Coherency** — top-K decode oracle on Qwen3.5-2B M2 Pro: slots 0 and 1 bit-identical between sequential and batched paths; slots 2 and 3 within fp16 noise (4/5 and 3/5 overlap with minor reordering — same pattern as M5 Max). No cross-slot corruption.

**Qwen3.5-9B-4bit** ran B=1/8/32 (decode 31/28/97 tok/s); B=64 thrashed on 32 GB unified memory (KV peak exceeded available, decode wall = 240 s). Expected ceiling for 9B at B=64 on this box, not a fix-related regression — same model loads cleanly at B=32.

**Comment by TheTom** [comment]
Fixed in 0.2.2
